File: sumtozero.py
from typing import List
def solve(a: List[int]) -> int:
    n = len(a)
    sum = 0
    maxLen = 0
    mp = {}
    for i in range(n):
        sum += a[i]
        if sum == 0:
            maxLen = i + 1
        elif sum in mp:
            maxLen = max(maxLen, i - mp[sum])
        else: mp[sum] = i
    return maxLen

if __name__ == "__main__":
    a = [9, -3, 3, -1, 6, -5]
    print("The length is ", solve(a))
    

File: sumtozero.cpp
#include <iostream>
#include <vector>
#include <unordered_map>
using namespace std;

int longestSubArray(vector<int> &arr)
{
    int n = arr.size();
    int max_len = 0;
    int sum = 0;
    unordered_map<int, int> sumIndexMap;

    for (int i = 0; i < n; i++)
    {
        sum += arr[i];
        if (sum == 0)
        {
            max_len = i + 1;
        }
        else if (sumIndexMap.find(sum) != sumIndexMap.end())
        {
            max_len = max(max_len, i - sumIndexMap[sum]);
        }
        else
        {
            sumIndexMap[sum] = i;
        }
    }
    return max_len;
}

int main()
{
    vector<int> a = {9, -3, 3, -1, 6, -5};
    cout << longestSubArray(a) << endl;
    return 0;
}